nerfstudio/exporter/tsdf_utils.py

```diff
@@ -0,0 +1,336 @@
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+import trimesh
+from rich.console import Console
+from skimage import measure
+from torchtyping import TensorType
+
+from nerfstudio.exporter.exporter_utils import Mesh, render_trajectory
+from nerfstudio.pipelines.base_pipeline import Pipeline
+
+CONSOLE = Console(width=120)
+
+
+@dataclass
+class TSDF:
+    """
+    Class for creating TSDFs.
+    """
+
+    voxel_coords: TensorType[3, "xdim", "ydim", "zdim"]
+    """Coordinates of each voxel in the TSDF."""
+    values: TensorType["xdim", "ydim", "zdim"]
+    """TSDF values for each voxel."""
+    weights: TensorType["xdim", "ydim", "zdim"]
+    """TSDF weights for each voxel."""
+    colors: TensorType["xdim", "ydim", "zdim", 3]
+    """TSDF colors for each voxel."""
+    voxel_size: TensorType[3]
+    """Size of each voxel in the TSDF. [x, y, z] size."""
+    origin: TensorType[3]
+    """Origin of the TSDF [xmin, ymin, zmin]."""
+    truncation_margin: float = 5.0
+    """Margin for truncation."""
+
+    def to(self, device: str):
+        """Move the tensors to the specified device.
+
+        Args:
+            device: The device to move the tensors to. E.g., "cuda:0" or "cpu".
+        """
+        self.voxel_coords = self.voxel_coords.to(device)
+        self.values = self.values.to(device)
+        self.weights = self.weights.to(device)
+        self.colors = self.colors.to(device)
+        self.voxel_size = self.voxel_size.to(device)
+        self.origin = self.origin.to(device)
+        return self
+
+    @property
+    def device(self):
+        """Returns the device that voxel_coords is on."""
+        return self.voxel_coords.device
+
+    @property
+    def truncation(self):
+        """Returns the truncation distance."""
+        # TODO: clean this up
+        truncation = self.voxel_size[0] * self.truncation_margin
+        return truncation
+
+    @staticmethod
+    def from_aabb(aabb: TensorType[2, 3], volume_dims: TensorType[3]):
+        """Returns an instance of TSDF from an axis-aligned bounding box and volume dimensions.
+
+        Args:
+            aabb: The axis-aligned bounding box with shape [[xmin, ymin, zmin], [xmax, ymax, zmax]].
+            volume_dims: The volume dimensions with shape [xdim, ydim, zdim].
+        """
+
+        origin = aabb[0]
+        voxel_size = (aabb[1] - aabb[0]) / volume_dims
+
+        # create the voxel coordinates
+        xdim = torch.arange(volume_dims[0])
+        ydim = torch.arange(volume_dims[1])
+        zdim = torch.arange(volume_dims[2])
+        grid = torch.stack(torch.meshgrid([xdim, ydim, zdim], indexing="ij"), dim=0)
+        voxel_coords = origin.view(3, 1, 1, 1) + grid * voxel_size.view(3, 1, 1, 1)
+
+        # initialize the values and weights
+        values = -torch.ones(volume_dims.tolist())
+        weights = torch.zeros(volume_dims.tolist())
+        colors = torch.zeros(volume_dims.tolist() + [3])
+
+        # TODO: move to device
```

Comment:
Why not initialize all these values on device?

Reply:
Possibly should be switched to that way.